File: frontend/src/types/availability.ts
export interface AvailabilityInterval {
  start: string;
  end: string;
  durationMinutes: number;
}

export interface DeviceAvailability {
  deviceId: number;
  deviceName: string;

  totalOccupiedMinutes: number;
  totalAvailableMinutes: number;

  occupied: AvailabilityInterval[];
  available: AvailabilityInterval[];
}

File: backend/app/schemas/availability.py
from datetime import datetime

from pydantic import BaseModel


class AvailabilityInterval(BaseModel):
    start: datetime
    end: datetime
    durationMinutes: int


class DeviceAvailability(BaseModel):
    deviceId: int
    deviceName: str

    totalOccupiedMinutes: int
    totalAvailableMinutes: int

    occupied: list[AvailabilityInterval]
    available: list[AvailabilityInterval]